src/OpenLoco/Graphics/Gfx.cpp

```diff
@@ -258,12 +258,91 @@ namespace OpenLoco::Gfx
     // 0x004957C4
     int16_t clipString(int16_t width, char* string)
     {
+        if (width < 6)
+        {
+            *string = '\0';
+            return 0;
+        }
 
-        registers regs;
-        regs.di = width;
-        regs.esi = (int32_t)string;
-        call(0x004957C4, regs);
-        return regs.cx;
+        // If width of the full string is less than allowed width then we don't need to clip
+        auto clippedWidth = getStringWidth(string);
+        if (clippedWidth <= width)
+        {
+            return clippedWidth;
+        }
+
+        // Append each character 1 by 1 with an ellipsis on the end until width is exceeded
+        std::string bestString;
+        std::string curString;
+
+        for (const auto* chr = string; *chr != '\0'; ++chr)
+        {
+            curString.push_back(*chr);
+            switch (*chr)
+            {
+                case ControlCodes::move_x:
+                    curString.push_back(*++chr);
+                    break;
+
+                case ControlCodes::adjust_palette:
+                case 3:
+                case 4:
+                    curString.push_back(*++chr);
+                    break;
+
+                case ControlCodes::newline:
+                case ControlCodes::newline_smaller:
+                case ControlCodes::font_small:
+                case ControlCodes::font_large:
+                case ControlCodes::font_bold:
+                case ControlCodes::font_regular:
+                case ControlCodes::outline:
+                case ControlCodes::outline_off:
+                case ControlCodes::window_colour_1:
+                case ControlCodes::window_colour_2:
+                case ControlCodes::window_colour_3:
+                case 0x10:
+                    break;
+
+                case ControlCodes::inline_sprite_str:
+                    curString.push_back(*++chr);
+                    curString.push_back(*++chr);
+                    curString.push_back(*++chr);
+                    curString.push_back(*++chr);
+                    break;
+
+                default:
+                    if (*chr <= 0x16)
+                    {
+                        curString.push_back(*++chr);
+                        curString.push_back(*++chr);
+                    }
+                    else if (*chr < 32)
+                    {
+                        curString.push_back(*++chr);
+                        curString.push_back(*++chr);
+                        curString.push_back(*++chr);
+                        curString.push_back(*++chr);
+                    }
```

Comment:
I'm not sure what format codes these are targeting. The only ones left in this range appear to be:
```
        // Arguments: int8_t, int8_t
        constexpr uint8_t newline_x_y = 17;

        // Arguments: uint32_t
        constexpr uint8_t inline_sprite_str = 23;
```

Moreover, there are some that have not been targeted at all: the large numeric arguments (`123 <= *chr <= 123+20`) and text colours (`144 <= *chr <= 157`)

Reply:
I copied the code from the string width code. Might still need some massaging.

Reply:
I looked into the other parts and they are not relevant as this is a post formatString string so all of the numeric argument control codes will already have been processed into normal control codes. Also the colours don't change the width so no special case is required.